### PR TITLE
feat(textarea): add labelPlacement

### PR DESCRIPTION
**Pull request by sean-perkins**
## Pull request checklist

Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
  - Some docs updates need to be made in the `ionic-docs` repo, in a separate PR. See the [contributing guide](https://github.com/ionic-team/ionic-framework/blob/main/.github/CONTRIBUTING.md#modifying-documentation) for details.
- [x] Build (`npm run build`) was run locally and any changes were pushed
- [x] Lint (`npm run lint`) has passed locally and any fixes were made for failures

## Pull request type

 

 

Please check the type of change your PR introduces:
- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe): 

## What is the current behavior?

With the new textarea control, developers need a way of placing the label similar to how they could using position on `ion-label`.

Issue URL: N/A

## What is the new behavior?

- Adds support for positioning the label at the start, end, floating, fixed, and stacked. Floating, fixed, and stacked have the same behavior as the position property on `ion-label`. Start and end should have similar behavior as slot on `ion-item` (though may look slightly different due to `ion-textarea` not having a bottom border by default).
- Fixes the appearance of `ion-textarea` across modes and label placement (text is consistently spaced from the top of the container, min height accounts for 3 rows of text)

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: core/src/components/textarea/test/label-placement/textarea.e2e.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+test.describe('textarea: label placement floating', () => {
+  test('label should appear above the textarea when there is a value', async ({ page }) => {
+    await page.setContent(`
+      <ion-textarea label="Email" value="example@ionic.io" label-placement="floating"></ion-textarea>
+    `);
+
+    const textarea = page.locator('ion-textarea');
+    expect(await textarea.screenshot()).toMatchSnapshot(
+      `input-placement-floating-value-${page.getSnapshotSettings()}.png`
```
Comment (sean-perkins) [inline core/src/components/textarea/test/label-placement/textarea.e2e.ts:68]: Need to:
- Update this to `textarea-`
- Remove reference screenshots that were captured under the wrong name
- Re-run to generate screenshots

File: core/src/components/textarea/textarea.scss [hunk 2]
```diff
@@ -283,3 +304,219 @@
...
+  @include transform-origin(start, top);
+
+  max-width: 100%;
+}
+
+/**
+ * This makes the label sit over the textarea
+ * when the textarea is blurred and has no value.
+ */
+:host(.textarea-label-placement-floating) .label-text-wrapper {
```
Comment (liamdebeasi) [inline core/src/components/textarea/textarea.scss:491]: The label alignment is odd when the label overlays the textarea:
image


I realize it is not supposed to be centered according to https://material-components.github.io/material-components-web-catalog/#/component/text-field

When comparing with the textareas in the link above, it looks like the text is positioned slightly higher than where the label sat (where this PR has it the other way around)

**Review by liamdebeasi — COMMENTED** [review]
I also noticed that there are no tests for an overflowing label. Are you planning on doing that when you do `fill`?

**Review by amandaejohnston — COMMENTED** [review]
I noticed the line height is ~1.5x taller currently, only allowing for 2 lines of text:
two lines

Compare this with the 7.x branch, which has a more standard line height and allows 3 lines:
three lines

Was this intentional? The PR description mentions it should be 3 lines. (It may be good to have a test with a multi-line value to account for this.)

**Comment by sean-perkins** [comment]
Sorry, this PR is in a bit of a "middle state". Trying to combine the feedback from Liam against the Material Design web component examples. In MD the line height of the textarea is ~28px. 

Textarea examples are at the bottom: https://material-components.github.io/material-components-web-catalog/#/component/text-field

Unfortunately these styles effect the textarea sizing and label placement (based on how we handle floating labels; rendering both containers and transforming the label over the textarea), so they either need to be accounted for or accomplished with alternative styles. I'm still exploring these challenges to see where we align vs. if we will have minor differences between our textarea and theirs. 

I can re-request reviews once I navigate these challenges.

**Comment by sean-perkins** [comment]
Ok, this is ready for another review. The modern rendering attempts to visually align with the legacy `ion-item` implementation (does not visually match MD web components 1:1).

I will create new tests/screenshots for multiple lines of text/text truncation in a follow-up PR.

**Review by liamdebeasi — COMMENTED** [review]
Looks like there are some legacy screenshot diffs. Maybe this needs to be synced with `feature-7.0` to account for https://github.com/ionic-team/ionic-framework/commit/eea91bbbe1a4ccc4797742ee4e2c320dba6d9517

Also the floating label placement does not look correct when compared with the MD spec:

| `ionic` | `spec` |
| - | - |
| <video src="https://user-images.githubusercontent.com/2721089/207388556-481fc6e1-7d75-487f-871c-0f0c9d6da5f9.mov"></video> | <video src="https://user-images.githubusercontent.com/2721089/207388623-5e3e2af5-97e5-4b22-9ffa-da24a6da9c02.mov"></video> |

(I know one has the outline and one does not, but I would expect the translations to be similar like they are on input)

**Comment by sean-perkins** [comment]
Closing this in favor of https://github.com/ionic-team/ionic-framework/pull/26492

**Pull request closed without merging**